src/runtime/halide_hexagon_dma_api.cpp

```diff
@@ -0,0 +1,261 @@
+void* halide_hexagon_dmaapp_get_memory(void* user_context, halide_buffer_t *roi_buf, \
+                                  bool padding, halide_hexagon_dma_user_fmt_t type) {
+    static void *pcache = NULL;     // ASSUMPTION: single frame only, lifetime scope is unknown will inner loop call
+
+    halide_assert(user_context, roi_buf != NULL);
+    //Check if roi_buf is already allocated
+    if(pcache != NULL) {
+        return pcache;
+    } else {
+        void *vret = NULL;
+        int nCircularFold;
+
+        halide_hexagon_dma_user_component_t comp;
+        if(halide_hexagon_dma_comp_get(user_context, roi_buf, comp)) {
+            error(user_context) << "Failure to get the component\n" ;
+            return NULL;
+        }
+        /////////////////////////////////////////////////////
+        //Problem with not including the inframe buf
+        // Have to take the global version of dma context
+        /////////////////////////////////////////////////////
+        HexagonDmaContext hexagon_dma_context(user_context);
+        halide_assert(user_context, hexagon_dma_context.get_context() != NULL);
+        p_dma_context dma_ctxt = hexagon_dma_context.get_context();
+
+        // ASSUMPTION: no folding
+        nCircularFold = 1;
+        ////////////////////////////////////////////////////////////
+        // Divide Frame to predefined tiles in Horizontal Direction
+        //fold_width = roi_buf->dim[0].extent;
+        // Divide Frame to predefined tiles in Vertical Direction
+        //fold_height = roi_buf->dim[1].extent;
+        // Each tile is again vertically split in to predefined DMA transfers
+        // Stride is aligned to Predefined Value
+        //fold_stride = roi_buf->dim[1].stride;
+        ///////////////////////////////////////////////////////////////
+     
+        /////////////////////////////////////////////////////////
+        //ASSUMPTION: Assuming paddng and fmt type to be default here since we really do  not have the inframe
+        //int halide_hexagon_dmart_get_padding(void* user_context, uintptr_t frame);
+        //int halide_hexagon_dmart_get_fmt_type(void* user_context, uintptr_t frame)
+        ///////////////////////////////////////////////////
+        // allocate L2$
+        vret = halide_hexagon_dma_memory_alloc(user_context, dma_ctxt, comp, roi_buf->dim[0].extent, \
+                         roi_buf->dim[1].extent, roi_buf->dim[1].stride, nCircularFold, padding, type);
+        if(vret == NULL) {
+            error(user_context) << "Failed to alloc host memeory." <<"\n";
+            return NULL;
+        }
+        pcache = vret;
+        return vret;
+    }
+}     
+
+
+//Step 1: Updates/Prepare the DMA for transfer
+//Step 2: Update ROI (to be transfered) information to the host
+//Step 3: copy to host: Actual transfer of Data
+//Step 4: Wait for DMA to be finished
+int halide_buffer_copy(void *user_context, halide_buffer_t *frame_buf, void *ptr, halide_buffer_t *roi_buf) {

```

Comment:
This already has an impleemntation: https://github.com/halide/Halide/blob/54dbb6b3f27d3be0c932945a1b7576c303725509/src/runtime/device_interface.cpp#L495

Your PR should not implement this function. Your PR should implement halide_hexagon_dma_copy_to_host, and the Halide implementation of halide_buffer_copy will call that.